### PR TITLE
Made fullscreen mode configurable from manifest.json.

### PR DESCRIPTION
**Pull request by niranjan-borawake**
This solves #34.

### CODE REVIEW

**Comment by niranjan-borawake** [comment]
Hi @collingreen, 
Did you have a look at this? 
Could  you please confirm if this is fine?  Or suggest any changes / improvements if required.

**Comment by collingreen** [comment]
What is the effect of adding FLAG_FORCE_NOT_FULLSCREEN? [The docs](http://developer.android.com/reference/android/view/WindowManager.LayoutParams.html) say:
> Window flag: override FLAG_FULLSCREEN and force the screen decorations (such as the status bar) to be shown.

What is the behavior without that flag? Is that a desired functionality we should also maintain if we are going to make this configurable?

**Comment by rampr** [comment]
I think this might disable the multitasking capability 



—ramSent from Mailbox

**Comment by rampr** [comment]
Sorry, spoke too soon, didn't realize this is for android



—ramSent from Mailbox

**Comment by niranjan-borawake** [comment]
@collingreen ,
`FLAG_FORCE_NOT_FULLSCREEN` makes sure the top notifications bar i.e. status bar is always visible while the app is running. If `FLAG_FULLSCREEN` is set then it is overridden. 

Without this flag the configuration defaults to the `FLAG_FULLSCREEN` which is current default full-screen mode.

If this config property is ignored in `manifest.json` then it defaults to `false` which is current default full-screen mode. Only when the config property is set to  `true` these code changes come into picture and disable full-screen mode making sure the notifications bar is visible.

**Comment by collingreen** [comment]
I understand your PR - I'm wondering what the default android behavior is when no flag is set. It sounds like there are three options - the default, forcing the status bar visible, and forcing the status bar hidden. Is that true? If so, it would be more complete to expose all three options instead of restricting to one or the other.

**Comment by niranjan-borawake** [comment]
@collingreen,
By default, Android shows status bar (non fullscreen) but devkit's current implementation hides status-bar ( as per the comment [here](https://github.com/niranjan-borawake/native-android/blob/master/TeaLeaf/src/com/tealeaf/TeaLeaf.java#L476) we are making the games full-screen by default but we want the status bar to be always visible in our game). Yes, I do agree that due to previous code changes we ended up with three options which was a bit confusing. 
I have made code changes which provide two options as below,
1) default Android behavior ( shows status bar, fullscreen  = false ) 
2) forcing the status bar to be hidden ( default behavior of devkit, hides status bar, fullscreen =  true ) 

If we don't set `fullscreen` property in `manifest.json` then game runs in fullscreen mode by default.

**Comment by aigninja** [comment]
Our game also wants statusbar to show server notifications.
@collingreen  could you please provide such feature, so we can show statusbar?

**Comment by collingreen** [comment]
@aigninja you should post feature requests as their own issues or on the google group

**Comment by niranjan-borawake** [comment]
@collingreen , is there a way this can be handled?

**Comment by collingreen** [comment]
@niranjan-borawake - The native side is being actively worked on right now by a different group but I can work on this with you as a temporary measure.

My first thought is that adding a top level manifest flag that has no effect on all the other platforms feels not ideal. iOS has a similar flag - it would be nice to combine this with a matching PR that sets the plist entry for ios builds as well (although this comes with serious ramifications, like making games ineligible for ipad multitasking). I don't know how the fullscreen api works in browsers these days, but if it makes sense an equivalent there would round out the 'fullscreen' feature. Alternatively, it could be added as an android-only flag by moving the flag into the android section and/or changing the name.

If we're being thorough, this should also include an update to the docs about available manifest flags.

Which direction do you think is best?

**Comment by niranjan-borawake** [comment]
@collingreen , I completely agree with all the points you have raised here.

Firstly, is there a way rather is it possible that, this is brought to notice of the group working on native and they take care of this? That way they would make sure it is handled for all other platforms. 

As a temporary fix I think the idea of adding it as a android-only flag would be best. I'm not much aware of the iOS environment and so am not at all the right person to comment on it. Being said this if it makes sense for iOS as well then top level flag would be the perfect one.

How about we decide on adding as android only flag for now and then once iOS counterpart is in we deprecate android-only flag?

And once this PR is ready to go in, I'll send out another one for docs as well.

**Comment by collingreen** [comment]
I've talked to the native group already about it - their take was to merge it here and they will figure out what the right move is on their end when it comes up.

I'm reluctant to add a flag that we know will be deprecated in short order, but maybe that's fine since there is no specific timeline for the next step. If you want to go that route, you should update your PR to put the fullscreen flag inside the android block of the manifest instead of the top level.

**Comment by niranjan-borawake** [comment]
Thanks @collingreen. 
I have updated PR to move `fullscreen` flag to android block.

**Comment by niranjan-borawake** [comment]
@collingreen , is this good to merge? Are there any other corrections/improvements to be done?

**Comment by collingreen** [comment]
It looks good - I'll test it when I'm back at my work machine next week and get it merged.

**Comment by collingreen** [comment]
Tested on a nexus 5 - looks good. Merging.

**Comment by niranjan-borawake** [comment]
Thanks a lot @collingreen. This will certainly help.

**Comment by collingreen** [comment]
Of course - thanks for contributing and for your patience!

File: build.js [hunk 1]
```diff
@@ -690,6 +690,7 @@ function updateManifest(api, app, config, opts) {
     version: "" + config.version,
     appid: app.manifest.appID.replace(PUNCTUATION_REGEX, ""), // Strip punctuation.,
     shortname: app.manifest.shortName,
+    fullscreen: app.manifest.android.fullscreen,
```
Comment (rampr) [inline build.js]: won't this break for games that haven't set this property in the manifest ?

**Comment by collingreen** [comment]
Thanks for having a look @rampr!

The rest of build.js also assumes `app.manifest.android` exists (I believe it gets a default value if it isn't set) so accessing `fullscreen` on that object should not cause any new problems there (it will be undefined, which is falsy). Some examples:

    if (app.manifest.android.resDir) {

    var splashPaths = app.manifest.android.splash || app.manifest.splash || DEFAULT_SPLASH_CONFIG;

    var android = app.manifest.android;
    var iconPath = android.icons && android.icons.alerts && android.icons.alerts[name];


Double checked to be sure by building a test project with some variations of config:

- Leaving the android block out of the manifest results in a successful build (fullscreen).
- Leaving the fullscreen out of the android block results in a successful build (fullscreen).
- Setting the android.fullscreen flag to true results in a successful build (fullscreen).
- Setting the android.fullscreen flag to false results in a successful build (NOT fullscreen).

**Comment by niranjan-borawake** [comment]
@rampr , thanks for your fresh pair of eyes on this.
@collingreen , thanks for double checking all above mentioned scenarios.

Even I had made sure I checked all these scenarios before raising a PR.
Also, I have raised another [PR](https://github.com/gameclosure/doc/pull/277) for updating documention for this. Could someone please have a look at that too.
Thanks again.

**Comment by niranjan-borawake** [comment]
@collingreen , @rampr ,
Any idea about when this would be released as part of new devkit-core release?

**Comment by jishnu7** [comment]
@martell is handling native side now, I think. you should as him.

**Comment by martell** [comment]
@niranjan-borawake As @jishnu7 said I am managing the native end for the moment.
To get my attention make sure you tag me in the post as sometimes github doesn't give me notifications but I get an email when you @ me.

landed in https://github.com/gameclosure/devkit-core/commit/c196e5e0a8f656b1fc91022f243ca24fb472be60

**Comment by collingreen** [comment]
Thanks @martell!

**Comment by niranjan-borawake** [comment]
Thanks @martell . Sure i'll tag you whenever required.

**Pull request merged**